docs/source/notes/rref.rst

```diff
@@ -0,0 +1,304 @@
+:ref:`distributed-rpc-framework` before proceeding.
+
+Background
+^^^^^^^^^^
+
+RRef stands for Remote REFerence. It is a reference of an object which locates
+on the local or a remote worker, and transparently handles reference counting
+under the hood. Conceptually, it can be considered as a distributed shared
+pointer. Applications can create an RRef by calling
+`torch.distributed.rpc.remote`. Each RRef is owned by the callee worker of the
+`torch.distributed.rpc.remote` call (i.e., owner) and can be used by multiple
+users. The owner stores the real data and keeps track of the global reference
+count. Every RRef can be uniquely identified by a global ``RRefId``,
+which is assigned at the time of creation on the caller of the
+`torch.distributed.rpc.remote` call.
+
+On the owner worker, there is only one ``OwnerRRef`` instance, which contains
+the real data, while on user workers, there can be as many ``UserRRefs`` as
+necessary, and ``UserRRef`` does not hold the data. All usage on the owner will
+retrieve the unique ``OwnerRRef`` instance using the globally unique ``RRefId``.
+A ``UserRRef`` will be created when it is used as an argument or return value in
+``rpc`` or ``remote`` invocation, and the owner will be notified according to
+update the reference count. An ``OwnerRRef`` and its data will be deleted when
+there is no ``UserRRef`` instances globally and there are no reference to the
+``OwnerRRef`` on the owner as well.
+
+
+Assumptions
+^^^^^^^^^^^
+
+RRef protocol is design with the following assumptions.
+
+- **Transient Network Failures**: The RRef design aims to handle transient
+  network failures by retrying messages. Node crashes or permanent network
+  partition is beyond the scope. When those incidents occur, the application
+  may take down all workers, revert to the previous checkpoint, and resume
+  training.
+- **Non-idempotent UDFs**: We assume UDFs are not idempotent and therefore
+  cannot be retried. However, internal RRef control messages will be made
+  idempotent and retryable.
+- **Out of Order Message Delivery**: We do not assume message delivery order
+  between any pair of nodes, because both sender and receiver are using multiple
+  threads. There is no guarantee on which message will be processed first.
+
+
+RRef Lifetime
+^^^^^^^^^^^^^
+
+The goal of the protocol is to delete an ``OwnerRRef`` at an appropriate time.
+The right time to delete an ``OwnerRRef`` is when there are no living
+``UserRRef`` instances and Python GC also agrees to delete the ``OwnerRRef``
+instance on the owner. The tricky part is to determine if there are any living
+``UserRRef`` instances.
+
+Design Reasoning
+----------------
+
+A user can get a ``UserRRef`` in three situations:
+
+1) Receiving a ``UserRRef`` from the owner.
+2) Receiving a ``UserRRef`` from another user.
+3) Creating a new ``UserRRef`` owned by another worker.
+
+
+Case 1 is the simplest where the owner initiates the fork, and hence it can
```

Comment:
Describe what "initiates the fork" means? I'm assuming this means they call `rpc.remote` or something similar.

Reply:
Yes, let me edit